lib/sshutils/networking/networking.go

```diff
@@ -0,0 +1,399 @@
+	NetworkingOperationListen Operation = "listen"
+	// NetworkingOperationListenAgent is used to create a local ssh-agent listener.
+	NetworkingOperationListenAgent Operation = "listen-agent"
+	// NetworkingOperationListenX11 is used to create a local x11 listener.
+	NetworkingOperationListenX11 Operation = "listen-x11"
+)
+
+// X11Config contains information used by the child process to set up X11 forwarding.
+type X11Request struct {
+	x11.ForwardRequestPayload
+	// DisplayOffset is the first display that we should try to get a unix socket for.
+	DisplayOffset int
+	// MaxDisplay is the last display that we should try to get a unix socket for, if all
+	// displays before it are taken.
+	MaxDisplay int
+}
+
+// NewProcess starts a new networking process with the given command, which should
+// be pre-configured from a ssh server context with Teleport reexec settings.
+func NewProcess(ctx context.Context, cmd *exec.Cmd) (*Process, error) {
+	// Create the socket to communicate over.
+	remoteConn, localConn, err := uds.NewSocketpair(uds.SocketTypeDatagram)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer remoteConn.Close()
+	remoteFD, err := remoteConn.File()
+	if err != nil {
+		localConn.Close()
+		return nil, trace.Wrap(err)
+	}
+	defer remoteFD.Close()
+	cmd.ExtraFiles = append(cmd.ExtraFiles, remoteFD)
+
+	// Propagate stderr from the spawned Teleport process to log any errors.
+	cmd.Stderr = os.Stderr
+
+	proc := &Process{
+		cmd:  cmd,
+		conn: localConn,
+		done: make(chan struct{}),
+	}
+
+	if err := proc.start(ctx); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return proc, nil
+}
+
+// start the the networking process.
+func (p *Process) start(ctx context.Context) error {
+	if err := p.cmd.Start(); err != nil {
+		p.conn.Close()
+		return trace.Wrap(err)
+	}
+
+	// The child process writes errors to the parent connection for logging purposes.
+	go func() {
+		for {
+			buf := make([]byte, RequestBufferSize)
+			n, err := p.conn.Read(buf)
+			if err != nil {
+				if utils.IsOKNetworkError(err) {
+					return
+				}
+				slog.WarnContext(ctx, "Failed to read error from networking process.", "error", err)
+				return
+			}
+
+			if n > 0 {
+				slog.WarnContext(ctx, "Received unexpected error from networking process.", "error", string(buf[:n]))
+			}
+		}
+	}()
+
+	go func() {
+		defer close(p.done)
+		defer p.conn.Close()
+		// Ensure unexpected cmd failures get logged.
+		if err := p.cmd.Wait(); err != nil && !p.killed.Load() {
+			slog.WarnContext(ctx, "Networking process exited early with unexpected error.", "error", err)
+		}
+	}()
+
+	return nil
+}
+
+// Close stops the process and frees up its related resources.
+func (p *Process) Close() error {
+	select {
+	case <-p.done:
+	default:
+		// We use an interrupt signal because the main control socket is not connection oriented,
+		// meaning reading from the child side won't unblock with the parent closes it.
+		p.cmd.Process.Signal(os.Interrupt)
```

Comment:
This is all fine and good until the parent dies without sending the signal - one of the good properties of the pipe trick (which we are 100% screwing up on the parent side by keeping both sides of the pipes in `srv.ServerContext` alive) is that when the other side that's holding the pipe dies for any reason, the pipe unblocks.